src/OrchardCore.Modules/OrchardCore.Media/Controllers/AdminController.cs

```diff
@@ -201,96 +201,91 @@ public async Task<IActionResult> Upload(string path, string extensions)
             (_, _, _) => Task.FromResult<IActionResult>(Ok(new { })),
             async (files) =>
             {
-                if (string.IsNullOrEmpty(path))
+                var result = await ProcessMediaUploadAsync(path, files, allowedExtensions);
+
+                return Ok(new { files = result.ToArray() });
+            });
+    }
+
+    private async Task<List<object>> ProcessMediaUploadAsync(string path, IEnumerable<IFormFile> files, HashSet<string> allowedExtensions)
+    {
+        if (string.IsNullOrEmpty(path))
+        {
+            path = string.Empty;
+        }
+
+        var result = new List<object>();
+
+        // Loop through each file in the request.
+        foreach (var file in files)
+        {
+            var extension = Path.GetExtension(file.FileName);
+
+            if (!allowedExtensions.Contains(extension))
+            {
+                result.Add(new
                 {
-                    path = string.Empty;
+                    name = file.FileName,
+                    size = file.Length,
+                    folder = path,
+                    error = S["This file extension is not allowed: {0}", extension].ToString(),
+                });
+
+                if (_logger.IsEnabled(LogLevel.Information))
+                {
+                    _logger.LogInformation("File extension not allowed: '{File}'", file.FileName);
                 }
 
-                var result = new List<object>();
+                continue;
+            }
+
+            var fileName = _mediaNameNormalizerService.NormalizeFileName(file.FileName);
 
-                // Loop through each file in the request.
-                foreach (var file in files)
+            Stream stream = null;
+            try
+            {
+                var mediaFilePath = _mediaFileStore.Combine(path, fileName);
+                stream = file.OpenReadStream();
+                mediaFilePath = await _mediaFileStore.CreateFileFromStreamAsync(mediaFilePath, stream);
+
+                var mediaFile = await _mediaFileStore.GetFileInfoAsync(mediaFilePath);
+
+                stream = null; // Must always reopen the resulting file
```

Comment:
You should not just set this to null, but change the `PreCacheRemoteMedia` method to not take a stream at all.